content/blog/2025-05-13-flat-bril/index.md

```diff
@@ -0,0 +1,170 @@
++++
+title = "A Flattened Representation for Bril"
+[[extra.authors]]
+name = "Ernest Ng"
+[[extra.authors]]
+name = "Katherine Wu"
+[[extra.authors]]
+name = "Samuel Breckenridge"
++++
+
+Typically, the way to implement an interpreter is to create an explicit AST data type, but this requires allocating AST nodes on the heap. Instead, as demonstrated in [this blogpost](https://www.cs.cornell.edu/~asampson/blog/flattening.html) by Adrian, we can flatten the AST into an array, i.e. pack all AST nodes into one single contiguous array, and refer to children in the AST using array indices (as opposed to pointers). This has enormous performance benefits!
+
+Since Bril is based on commands rather than expressions, there are no ASTs to flatten. However, in the existing Rust interpreter, programs, functions and instructions are all represented using structs that contain pointers to the heap. For instance this is the representation of a value instruction:
+
+```Rust
+Value {
+        args: Vec<String>,
+        dest: String,
+        funcs: Vec<String>,
+        labels: Vec<String>,
+        op: ValueOps,
+        pos: Option<Position>,
+        op_type: Type,
+    }
+```
+
+`Vec`s and `String`s in Rust are both implemented using pointers, so the representation is clearly not flat even before considering that functions contain pointers to instructions and programs contain pointers to functions. Our goal was to adopt the same approach of flattening data structures and packing them into a single contiguous array for Bril. The resulting representation of Bril programs is then a flat file containing these arrays that can be directly mapped into memory and interpreted, analogous to the approach laid out in [another blogpost](https://www.cs.cornell.edu/~asampson/blog/flatgfa.html) by Adrian. This requires two main components:
+
+1. Infrastructure to convert existing Bril JSON files to/from our flattened format
+2. An alternate Bril interpreter that operates directly on the flattened data structure (as opposed to the existing one brili, which has to parse JSON)
+
+Our complete implementation can be found at https://github.com/ngernest/flat-bril/tree/main
```

Comment:
Consider integrating this link into the text, like "our complete implantation is [open source on GitHub][…]".